File: services/sabwa-node/src/db/scheduled.ts
/**
 * `sabwa_scheduled` repository — one-off and recurring outbound messages.
 *
 * Mirrors `services/sabwa-engine/src/db/scheduled.rs` so the Node engine and
 * the Rust engine remain swappable behind the same HTTP contract.
 *
 * Document shape (camelCase, matches the `ScheduledRow` DTO exposed by
 * `routes/scheduled.ts`):
 *
 *   {
 *     _id:           string            // "sch_<uuid>"
 *     projectId?:    string
 *     sessionId:     string
 *     kind:          'one_off' | 'recurring'
 *     scheduledFor:  Date              // UTC instant we should fire at
 *     cron?:         string            // only on `recurring`
 *     timezone?:     string            // IANA zone, defaults to 'UTC'
 *     targets:       { jid, type }[]
 *     payload:       JsonValue         // raw send payload (text/media/...)
 *     status:        'pending'|'sent'|'failed'|'cancelled'
 *     attemptCount:  number
 *     lastError?:    string
 *     sentAt?:       Date
 *     // recurring-parent bookkeeping:
 *     parentId?:     string            // set on materialised child rows
 *     lastFiredAt?:  Date              // set on the parent each time we
 *                                      // materialise the next instance
 *     createdAt:     Date
 *     updatedAt:     Date
 *   }
 *
 * The `findDue` query is the hot path — every tick the scheduler runs
 *   { status: 'pending', scheduledFor: { $lte: now } }
 * against this collection. A compound index on `(status, scheduledFor)` is
 * created lazily on first call.
 */

import { type Db, type Collection } from 'mongodb';

export const COLLECTION = 'sabwa_scheduled';

export type ScheduledKind = 'one_off' | 'recurring';
export type ScheduledStatus = 'pending' | 'sent' | 'failed' | 'cancelled';
export type ScheduledTargetType = 'individual' | 'group' | 'broadcast';

export interface ScheduledTarget {
  jid: string;
  /** Stored under `type` to match the Rust engine's wire contract. */
  type: ScheduledTargetType;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type JsonValue = any;

export interface ScheduledDoc {
  _id: string;
  projectId?: string;
  sessionId: string;
  kind: ScheduledKind;
  scheduledFor: Date;
  cron?: string;
  timezone?: string;
  targets: ScheduledTarget[];
  payload: JsonValue;
  status: ScheduledStatus;
  attemptCount: number;
  lastError?: string;
  sentAt?: Date;
  parentId?: string;
  lastFiredAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

/** Wire shape returned by `GET /v1/scheduled` — `id` instead of `_id`. */
export interface ScheduledRow {
  id: string;
  sessionId: string;
  kind: ScheduledKind;
  scheduledFor: Date;
  cron: string | null;
  timezone: string | null;
  status: ScheduledStatus;
  attemptCount: number;
  lastError: string | null;
  sentAt: Date | null;
  targets: ScheduledTarget[];
  payload: JsonValue;
}

function collection(db: Db): Collection<ScheduledDoc> {
  return db.collection<ScheduledDoc>(COLLECTION);
}

export function toRow(doc: ScheduledDoc): ScheduledRow {
  return {
    id: doc._id,
    sessionId: doc.sessionId,
    kind: doc.kind,
    scheduledFor: doc.scheduledFor,
    cron: doc.cron ?? null,
    timezone: doc.timezone ?? null,
    status: doc.status,
    attemptCount: doc.attemptCount ?? 0,
    lastError: doc.lastError ?? null,
    sentAt: doc.sentAt ?? null,
    targets: doc.targets ?? [],
    payload: doc.payload ?? null,
  };
}

let indexEnsured = false;
async function ensureIndexes(db: Db): Promise<void> {
  if (indexEnsured) return;
  indexEnsured = true;
  try {
    await collection(db).createIndex(
      { status: 1, scheduledFor: 1 },
      { name: 'status_scheduledFor' },
    );
    await collection(db).createIndex(
      { sessionId: 1, scheduledFor: 1 },
      { name: 'sessionId_scheduledFor' },
    );
  } catch {
    // index creation is best-effort; the queries still work without it.
  }
}

export interface InsertParams {
  _id: string;
  projectId?: string;
  sessionId: string;
  kind: ScheduledKind;
  scheduledFor: Date;
  cron?: string;
  timezone?: string;
  targets: ScheduledTarget[];
  payload: JsonValue;
  /** Set when this row is a materialised child of a recurring parent. */
  parentId?: string;
}

/** Insert a new pending row. */
export async function insert(db: Db, params: InsertParams): Promise<ScheduledDoc> {
  await ensureIndexes(db);
  const now = new Date();
  const doc: ScheduledDoc = {
    _id: params._id,
    projectId: params.projectId,
    sessionId: params.sessionId,
    kind: params.kind,
    scheduledFor: params.scheduledFor,
    cron: params.cron,
    timezone: params.timezone ?? 'UTC',
    targets: params.targets,
    payload: params.payload,
    status: 'pending',
    attemptCount: 0,
    parentId: params.parentId,
    createdAt: now,
    updatedAt: now,
  };
  await collection(db).insertOne(doc);
  return doc;
}

/**
 * Find all `pending` rows whose `scheduledFor` is in the past, oldest first.
 * `limit` caps the batch the worker drains per tick.
 *
 * Recurring parent rows are **excluded** from this query — they materialise
 * children via `next-instance` insertion rather than firing themselves. The
 * children carry `kind: 'one_off'` so they drop into this query naturally.
 */
export async function findDue(db: Db, now: Date, limit = 100): Promise<ScheduledDoc[]> {
  await ensureIndexes(db);
  return collection(db)
    .find({
      status: 'pending',
      scheduledFor: { $lte: now },
      kind: { $ne: 'recurring' },
    })
    .sort({ scheduledFor: 1 })
    .limit(Math.max(1, limit))
    .toArray();
}

/** Find all `pending` recurring parents that are eligible for materialisation. */
export async function findRecurringParents(db: Db): Promise<ScheduledDoc[]> {
  await ensureIndexes(db);
  return collection(db)
    .find({ status: 'pending', kind: 'recurring', cron: { $exists: true } })
    .toArray();
}

export async function findById(db: Db, id: string): Promise<ScheduledDoc | null> {
  return collection(db).findOne({ _id: id });
}

/** Flip status → `sent` and stamp `sentAt`. */
export async function markSent(db: Db, id: string): Promise<void> {
  const now = new Date();
  await collection(db).updateOne(
    { _id: id },
    {
      $set: { status: 'sent', sentAt: now, updatedAt: now, lastError: undefined },
      $inc: { attemptCount: 1 },
    },
  );
}

/** Flip status → `failed`, persist the error message, bump `attemptCount`. */
export async function markFailed(db: Db, id: string, error: string): Promise<void> {
  const now = new Date();
  await collection(db).updateOne(
    { _id: id },
    {
      $set: { status: 'failed', lastError: error, updatedAt: now },
      $inc: { attemptCount: 1 },
    },
  );
}

export interface UpdateParams {
  scheduledFor?: Date;
  cron?: string;
  timezone?: string;
  targets?: ScheduledTarget[];
  payload?: JsonValue;
}

/** Apply a patch. Returns `true` if a row was modified. */
export async function update(db: Db, id: string, patch: UpdateParams): Promise<boolean> {
  const set: Record<string, unknown> = { updatedAt: new Date() };
  if (patch.scheduledFor !== undefined) set.scheduledFor = patch.scheduledFor;
  if (patch.cron !== undefined) set.cron = patch.cron;
  if (patch.timezone !== undefined) set.timezone = patch.timezone;
  if (patch.targets !== undefined) set.targets = patch.targets;
  if (patch.payload !== undefined) set.payload = patch.payload;
  const res = await collection(db).updateOne(
    { _id: id, status: 'pending' },
    { $set: set },
  );
  return res.modifiedCount > 0;
}

/** Cancel a pending row (idempotent — non-pending rows are left alone). */
export async function cancel(db: Db, id: string): Promise<boolean> {
  const res = await collection(db).updateOne(
    { _id: id, status: 'pending' },
    { $set: { status: 'cancelled', updatedAt: new Date() } },
  );
  return res.modifiedCount > 0;
}

/** List rows for a session, optionally filtered by status. */
export async function listBySession(
  db: Db,
  sessionId: string,
  status?: ScheduledStatus,
): Promise<ScheduledDoc[]> {
  await ensureIndexes(db);
  const filter: Record<string, unknown> = { sessionId };
  if (status) filter.status = status;
  return collection(db).find(filter).sort({ scheduledFor: 1 }).toArray();
}

/** Stamp `lastFiredAt` on a recurring parent after materialising a child. */
export async function setLastFiredAt(db: Db, id: string, when: Date): Promise<void> {
  await collection(db).updateOne(
    { _id: id },
    { $set: { lastFiredAt: when, updatedAt: new Date() } },
  );
}

/** Attach an error message to a row without changing its status. */
export async function setLastError(db: Db, id: string, error: string): Promise<void> {
  await collection(db).updateOne(
    { _id: id },
    { $set: { lastError: error, updatedAt: new Date() } },
  );
}

/** Push a recurring parent forward to the next computed fire time. */
export async function advanceRecurringParent(
  db: Db,
  id: string,
  nextFireAt: Date,
): Promise<void> {
  await collection(db).updateOne(
    { _id: id },
    { $set: { scheduledFor: nextFireAt, updatedAt: new Date() } },
  );
}

/** Retire a recurring parent (e.g. its cron no longer yields future fires). */
export async function retireRecurringParent(db: Db, id: string): Promise<void> {
  await collection(db).updateOne(
    { _id: id },
    { $set: { status: 'sent', updatedAt: new Date() } },
  );
}

File: services/sabwa-engine/src/db/scheduled.rs
//! `sabwa_scheduled` — scheduled / recurring outbound messages.
//!
//! See SABWA_PLAN.md §3 "sabwa_scheduled".

use anyhow::{Context, Result};
use bson::{doc, oid::ObjectId, Bson};
use chrono::{DateTime, Utc};
use futures::TryStreamExt;
use mongodb::{Collection, Database};
use serde::{Deserialize, Serialize};

use crate::db::serde_dates::{chrono_dt, chrono_dt_opt};

pub const COLLECTION: &str = "sabwa_scheduled";

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ScheduledKind {
    OneOff,
    Recurring,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ScheduledStatus {
    Pending,
    Sent,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ScheduledTargetType {
    Individual,
    Group,
    Broadcast,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledTarget {
    pub jid: String,
    #[serde(rename = "type")]
    pub target_type: ScheduledTargetType,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledPayload {
    #[serde(rename = "type")]
    pub payload_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub body: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media_sab_file_id: Option<ObjectId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub caption: Option<String>,
    #[serde(default)]
    pub mention_all: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SabwaScheduled {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub project_id: ObjectId,
    pub session_id: ObjectId,
    pub kind: ScheduledKind,
    #[serde(with = "chrono_dt")]
    pub scheduled_for: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cron: Option<String>,
    pub timezone: String,
    pub targets: Vec<ScheduledTarget>,
    pub payload: ScheduledPayload,
    pub status: ScheduledStatus,
    #[serde(default)]
    pub attempt_count: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_error: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none", with = "chrono_dt_opt")]
    pub sent_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bull_job_id: Option<String>,
}

pub struct ScheduledRepo<'a> {
    col: Collection<SabwaScheduled>,
    _phantom: std::marker::PhantomData<&'a ()>,
}

impl<'a> ScheduledRepo<'a> {
    pub fn new(db: &'a Database) -> Self {
        Self {
            col: db.collection::<SabwaScheduled>(COLLECTION),
            _phantom: std::marker::PhantomData,
        }
    }

    pub async fn insert(&self, item: &SabwaScheduled) -> Result<ObjectId> {
        let res = self
            .col
            .insert_one(item)
            .await
            .context("sabwa_scheduled.insert")?;
        res.inserted_id
            .as_object_id()
            .context("inserted_id was not ObjectId")
    }

    /// Find all pending scheduled items whose fire time is in the past.
    /// Sorted oldest-first so the worker drains them in order. `limit` caps
    /// the batch size per tick.
    pub async fn find_due(&self, now: DateTime<Utc>, limit: i64) -> Result<Vec<SabwaScheduled>> {
        let cursor = self
            .col
            .find(doc! {
                "status": "pending",
                "scheduledFor": { "$lte": Bson::DateTime(now.into()) },
            })
            .sort(doc! { "scheduledFor": 1 })
            .limit(limit.max(1))
            .await
            .context("sabwa_scheduled.find_due")?;
        let out: Vec<SabwaScheduled> = cursor.try_collect().await.context("collect due")?;
        Ok(out)
    }

    pub async fn update_status(
        &self,
        id: &ObjectId,
        status: ScheduledStatus,
        last_error: Option<&str>,
    ) -> Result<()> {
        let status_bson = bson::to_bson(&status).context("encode ScheduledStatus")?;
        let mut set = doc! { "status": status_bson };
        if matches!(status, ScheduledStatus::Sent) {
            set.insert("sentAt", Bson::DateTime(bson::DateTime::now()));
        }
        if let Some(err) = last_error {
            set.insert("lastError", err);
        }
        self.col
            .update_one(doc! { "_id": id }, doc! { "$set": set, "$inc": { "attemptCount": 1 } })
            .await
            .context("sabwa_scheduled.update_status")?;
        Ok(())
    }

    pub async fn cancel(&self, id: &ObjectId) -> Result<()> {
        self.col
            .update_one(
                doc! { "_id": id, "status": "pending" },
                doc! { "$set": { "status": "cancelled" } },
            )
            .await
            .context("sabwa_scheduled.cancel")?;
        Ok(())
    }

    pub async fn find_by_session(
        &self,
        session_id: &ObjectId,
    ) -> Result<Vec<SabwaScheduled>> {
        let cursor = self
            .col
            .find(doc! { "sessionId": session_id })
            .sort(doc! { "scheduledFor": 1 })
            .await
            .context("sabwa_scheduled.find_by_session")?;
        let out: Vec<SabwaScheduled> = cursor.try_collect().await.context("collect scheduled")?;
        Ok(out)
    }
}

pub fn collection(db: &Database) -> Collection<SabwaScheduled> {
    db.collection::<SabwaScheduled>(COLLECTION)
}

// ---------------------------------------------------------------------------
// Phase 1 route-compat shims.
//
// Route handlers store string-form ids (e.g. `sch_<uuid>`) on the side and
// use simple flat row projections. We persist these via a fallback `Document`
// collection so the Repo<SabwaScheduled> typed path can stay strict.
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledTargetRow {
    pub jid: String,
    pub kind: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScheduledRow {
    pub id: String,
    pub session_id: String,
    pub kind: String,
    #[serde(with = "chrono_dt")]
    pub scheduled_for: DateTime<Utc>,
    pub cron: Option<String>,
    pub timezone: Option<String>,
    pub status: String,
    pub attempt_count: u32,
    pub last_error: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none", with = "chrono_dt_opt")]
    pub sent_at: Option<DateTime<Utc>>,
    pub targets: Vec<ScheduledTargetRow>,
    pub payload: serde_json::Value,
}

fn scheduled_doc_collection(db: &Database) -> Collection<bson::Document> {
    db.collection::<bson::Document>(COLLECTION)
}

pub async fn insert(
    db: &Database,
    scheduled_id: &str,
    project_id: &str,
    session_id: &str,
    kind: &str,
    scheduled_for: DateTime<Utc>,
    cron: Option<&str>,
    timezone: Option<&str>,
    targets: &[serde_json::Value],
    payload: &serde_json::Value,
) -> Result<()> {
    let col = scheduled_doc_collection(db);
    let targets_bson: Vec<Bson> = targets
        .iter()
        .map(|t| bson::to_bson(t).unwrap_or(Bson::Null))
        .collect();
    let payload_bson = bson::to_bson(payload).unwrap_or(Bson::Null);
    let doc = doc! {
        "_id": scheduled_id,
        "projectId": project_id,
        "sessionId": session_id,
        "kind": kind,
        "scheduledFor": Bson::DateTime(scheduled_for.into()),
        "cron": cron,
        "timezone": timezone,
        "targets": targets_bson,
        "payload": payload_bson,
        "status": "pending",
        "attemptCount": 0i32,
    };
    col.insert_one(doc).await.context("sabwa_scheduled.insert (raw)")?;
    Ok(())
}

pub async fn set_bull_job_id(db: &Database, scheduled_id: &str, job_id: &str) -> Result<()> {
    let col = scheduled_doc_collection(db);
    col.update_one(
        doc! { "_id": scheduled_id },
        doc! { "$set": { "bullJobId": job_id } },
    )
    .await
    .context("sabwa_scheduled.set_bull_job_id")?;
    Ok(())
}

pub async fn list(
    db: &Database,
    session_id: &str,
    status: Option<&str>,
) -> Result<Vec<ScheduledRow>> {
    let col = scheduled_doc_collection(db);
    let mut filter = doc! { "sessionId": session_id };
    if let Some(s) = status {
        filter.insert("status", s);
    }
    let cursor = col
        .find(filter)
        .sort(doc! { "scheduledFor": 1 })
        .await
        .context("sabwa_scheduled.list")?;
    let docs: Vec<bson::Document> =
        cursor.try_collect().await.context("collect scheduled docs")?;
    Ok(docs.into_iter().filter_map(doc_to_row).collect())
}

fn doc_to_row(d: bson::Document) -> Option<ScheduledRow> {
    let id = match d.get("_id") {
        Some(Bson::String(s)) => s.clone(),
        Some(Bson::ObjectId(o)) => o.to_hex(),
        _ => return None,
    };
    let session_id = match d.get("sessionId") {
        Some(Bson::String(s)) => s.clone(),
        Some(Bson::ObjectId(o)) => o.to_hex(),
        _ => String::new(),
    };
    let kind = d.get_str("kind").unwrap_or("one_off").to_string();
    let scheduled_for = match d.get("scheduledFor") {
        Some(Bson::DateTime(dt)) => dt.to_chrono(),
        _ => Utc::now(),
    };
    let cron = d.get_str("cron").ok().map(str::to_string);
    let timezone = d.get_str("timezone").ok().map(str::to_string);
    let status = d.get_str("status").unwrap_or("pending").to_string();
    let attempt_count = d.get_i32("attemptCount").unwrap_or(0).max(0) as u32;
    let last_error = d.get_str("lastError").ok().map(str::to_string);
    let sent_at = match d.get("sentAt") {
        Some(Bson::DateTime(dt)) => Some(dt.to_chrono()),
        _ => None,
    };
    let targets: Vec<ScheduledTargetRow> = match d.get("targets") {
        Some(Bson::Array(arr)) => arr
            .iter()
            .filter_map(|b| {
                let doc = b.as_document()?;
                Some(ScheduledTargetRow {
                    jid: doc.get_str("jid").unwrap_or("").to_string(),
                    kind: doc.get_str("type").unwrap_or("individual").to_string(),
                })
            })
            .collect(),
        _ => Vec::new(),
    };
    let payload = d
        .get("payload")
        .cloned()
        .map(|b| b.into_relaxed_extjson())
        .unwrap_or(serde_json::Value::Null);
    Some(ScheduledRow {
        id,
        session_id,
        kind,
        scheduled_for,
        cron,
        timezone,
        status,
        attempt_count,
        last_error,
        sent_at,
        targets,
        payload,
    })
}

pub async fn update(
    db: &Database,
    scheduled_id: &str,
    scheduled_for: Option<DateTime<Utc>>,
    cron: Option<&str>,
    timezone: Option<&str>,
    payload: Option<&serde_json::Value>,
) -> Result<()> {
    let col = scheduled_doc_collection(db);
    let mut set = bson::Document::new();
    if let Some(s) = scheduled_for {
        set.insert("scheduledFor", Bson::DateTime(s.into()));
    }
    if let Some(c) = cron {
        set.insert("cron", c);
    }
    if let Some(t) = timezone {
        set.insert("timezone", t);
    }
    if let Some(p) = payload {
        set.insert("payload", bson::to_bson(p).unwrap_or(Bson::Null));
    }
    if set.is_empty() {
        return Ok(());
    }
    col.update_one(doc! { "_id": scheduled_id }, doc! { "$set": set })
        .await
        .context("sabwa_scheduled.update (raw)")?;
    Ok(())
}

pub async fn cancel(db: &Database, scheduled_id: &str) -> Result<()> {
    let col = scheduled_doc_collection(db);
    col.update_one(
        doc! { "_id": scheduled_id, "status": "pending" },
        doc! { "$set": { "status": "cancelled" } },
    )
    .await
    .context("sabwa_scheduled.cancel (raw)")?;
    Ok(())
}
